source/docs/software/commandbased/commands-v3/structuring-v3-project.rst

```diff
@@ -0,0 +1,529 @@
+
+For a comprehensive guide to building a hierarchical testing system with result reporting and range verification, see :ref:`docs/software/commandbased/commands-v3/testing-commands-v3:Testing Commands v3 Projects`.
+
+### Quick Example
+
+.. code-block:: java
+
+   package frc.robot.commands;
+
+   import org.wpilib.commands3.Command;
+   import frc.robot.mechanisms.Drivetrain;
+   import static edu.wpi.first.units.Units.Seconds;
+
+   public class TestCommands {
+
+     public static Command testDrivetrain(Drivetrain drivetrain) {
+       return drivetrain.run(coroutine -> {
+         System.out.println("=== Testing Drivetrain ===");
+
+         // Test forward
+         drivetrain.tank(0.3, 0.3);
+         coroutine.wait(Seconds.of(1.0));
+         drivetrain.stop();
+
+         double leftDist = drivetrain.getLeftDistance();
+         double rightDist = drivetrain.getRightDistance();
+
+         // Verify readings are in expected range
+         if (leftDist < 0.1 || leftDist > 3.0) {
+           System.err.println("❌ Left encoder reading unusual: " + leftDist + "m");
+           return;
+         }
+
+         System.out.println("✓ Drivetrain test passed");
+       }).named("Test Drivetrain");
+     }
+   }
+
+## Common Mistakes to Avoid
+
+1. **Scheduling commands inside mechanisms**: Return commands from methods, don't schedule them
+
+.. code-block:: java
+
+   // ❌ BAD: Scheduling inside mechanism
+   public void grab() {
+     Scheduler.getDefault().schedule(run(coro -> { /* ... */ });
+   }
+
+   // ✅ GOOD: Return a command
+   public Command grab() {
+     return run(coro -> { /* ... */ }).named("Grab");
+   }
+
+2. **Forgetting to call yield()**: Always yield inside loops. A compiler plugin is in development to detect this.
+
+.. code-block:: java
+
+   // ❌ BAD: No yield
+   while (condition) {
+     doWork();
+   }
+
+   // ✅ GOOD: Yields every cycle
+   while (condition) {
+     doWork();
+     coroutine.yield();
+   }
+
+3. **Not naming commands**: All commands need names
+
+.. code-block:: java
+
+   // ❌ BAD: This is a compilation error - run() returns a builder
+   Command cmd = mechanism.run(coro -> { /* ... */ });
+
+   // ✅ GOOD: Has name (and compiles)
+   Command cmd = mechanism.run(coro -> { /* ... */ }).named("Action");
+
+4. **Exposing hardware control methods**: Only commands and sensor-reading methods should be public
+
+.. code-block:: java
+
+   // ❌ BAD: Public hardware control methods
+   public void setSpeed(double speed) {
+     motor.set(speed);
+   }
+
+   // ✅ GOOD: Private hardware control, public commands
+   private void setSpeed(double speed) {
+     motor.set(speed);
+   }
+
+   public Command grab() {
+     return run(coro -> { /* ... */ }).named("Grab");
+   }
+
+5. **Complex logic in RobotContainer**: Move complex commands to separate files. RobotContainer is not recommended in v3.
+
+## Example: Complete Small Project
+
+See the full example structure:
```

Comment:
This is confusing. Is it missing a full project or link to full project? or is it duplicative of the first section?